src/reader/slice_reader.rs

```diff
@@ -275,8 +275,8 @@ impl<'a> XmlSource<'a, ()> for &'a [u8] {
         }
     }
 
-    fn read_pi(&mut self, _buf: (), position: &mut usize) -> Result<&'a [u8]> {
-        let mut parser = PiParser::default();
+    fn read<P: Parser>(&mut self, _buf: (), position: &mut usize) -> Result<&'a [u8]> {
```

Comment:
`read<P: Parser>()` seems like an over-general name - how will this type of pattern work with `TextEvent`s?  Do you plan on a `TextParser`? 

If it will be used only for a couple specific circumstances, then `read_element<P: Parser>()` seems more appropriate.

Reply:
I do not known yet is there will be `TextParser` or not. I use "parser" structs to store state which is required to correctly understand where the corresponding event ends. Right now when we parse text there is no need to store any state, the first `<` starts markup. This probably will be changed, because entity references (`&something;`) can insert XML nodes into the text (that is why they are called "parsed entities" in the specification, because the result of their resolution is a text which is parsed into XML node) and it seems that for correct parsing we should resolve them immediately when their are encountered.

For now this is a my experimentation to modularize parser to reuse parts of them in DTD parsing. So the selected name could be a bit random, although I think, I can name it `read_with` which would even look more good from grammatical point of view.